cgroups v2 Support

Work needs to be done to the cgroups lib and containerd metrics interfaces to support cgroups v2 support.  With much of the work in 5.3 kernels this should be reasonable to start supporting as a first class feature and can be a replacement for v1 for some users.

runc recently gained support for v2 as well as crun.

---
(below are edited by maintainers)

## runc checklist
See https://github.com/opencontainers/runc/issues/2315

## containerd checklist
- [x] base https://github.com/containerd/containerd/pull/3799
- [x] PID metrics https://github.com/containerd/containerd/pull/3799
- [x] CPU metrics
- [x] Memory metrics
- [x] OOM handler: https://github.com/containerd/containerd/pull/4273
- [x] IO metrics: https://github.com/containerd/containerd/pull/3919

## containerd/CRI checklist
- [x]  metrics: https://github.com/containerd/cri/pull/1376
- [x] [enable cgroup namespace by default](https://github.com/containers/libpod/issues/4363): https://github.com/containerd/cri/pull/1371
- enable cgroup namespace for pod sandboxes as well as for containers? (To be determined)
- [x] disable cgroup namespace for privileged pods (discussed in https://github.com/kubernetes/enhancements/pull/1370): https://github.com/containerd/cri/pull/1415
- systemd cgroup driver for rootless (seems working without additional work: https://github.com/AkihiroSuda/critest-rootless-cgroup2)

## CI
- [x] CI (crun): https://github.com/containerd/containerd/pull/4279
- [X] CI (runc): https://github.com/containerd/containerd/pull/4408
- CI for rootless mode with systemd (will be a follow-up PR)

Seems we can't remove the prefix :cry:

```console
$ git diff upstream/master 
diff --git a/metrics/cgroups/v2/pids.go b/metrics/cgroups/v2/pids.go
index 6b693f65..704f87ec 100644
--- a/metrics/cgroups/v2/pids.go
+++ b/metrics/cgroups/v2/pids.go
@@ -26,7 +26,7 @@ import (
 
 var pidMetrics = []*metric{
        {
-               name: "pids_v2",
+               name: "pids",
                help: "The limit to the number of pids allowed",
                unit: metrics.Unit("limit"),
                vt:   prometheus.GaugeValue,
@@ -42,7 +42,7 @@ var pidMetrics = []*metric{
                },
        },
        {
-               name: "pids_v2",
+               name: "pids",
                help: "The current number of pids",
                unit: metrics.Unit("current"),
                vt:   prometheus.GaugeValue,

$ sudo containerd
INFO[2019-12-13T15:19:04.250646565+09:00] starting containerd                           revision=30c855b7ebe55e47520bfafa91d97a5e3ab8787f.m version=v1.3.0-173-g30c855b7.m
INFO[2019-12-13T15:19:04.292347976+09:00] loading plugin "io.containerd.content.v1.content"...  type=io.containerd.content.v1
INFO[2019-12-13T15:19:04.292417950+09:00] loading plugin "io.containerd.snapshotter.v1.btrfs"...  type=io.containerd.snapshotter.v1
INFO[2019-12-13T15:19:04.292619783+09:00] skip loading plugin "io.containerd.snapshotter.v1.btrfs"...  error="path /var/lib/containerd/io.containerd.snapshotter.v1.btrfs (ext4) must be a btrfs filesystem to be used with the btrfs snapshotter: skip plugin" type=io.containerd.sna
pshotter.v1
INFO[2019-12-13T15:19:04.292637575+09:00] loading plugin "io.containerd.snapshotter.v1.devmapper"...  type=io.containerd.snapshotter.v1
WARN[2019-12-13T15:19:04.292651194+09:00] failed to load plugin io.containerd.snapshotter.v1.devmapper  error="devmapper not configured"
INFO[2019-12-13T15:19:04.292659193+09:00] loading plugin "io.containerd.snapshotter.v1.aufs"...  type=io.containerd.snapshotter.v1
INFO[2019-12-13T15:19:04.295739560+09:00] loading plugin "io.containerd.snapshotter.v1.native"...  type=io.containerd.snapshotter.v1
INFO[2019-12-13T15:19:04.296165681+09:00] loading plugin "io.containerd.snapshotter.v1.overlayfs"...  type=io.containerd.snapshotter.v1
INFO[2019-12-13T15:19:04.296695724+09:00] loading plugin "io.containerd.snapshotter.v1.zfs"...  type=io.containerd.snapshotter.v1
INFO[2019-12-13T15:19:04.301592785+09:00] loading plugin "io.containerd.metadata.v1.bolt"...  type=io.containerd.metadata.v1
WARN[2019-12-13T15:19:04.301739188+09:00] could not use snapshotter devmapper in metadata plugin  error="devmapper not configured"
INFO[2019-12-13T15:19:04.301763508+09:00] metadata content store policy set             policy=shared
INFO[2019-12-13T15:19:04.302004419+09:00] loading plugin "io.containerd.differ.v1.walking"...  type=io.containerd.differ.v1
INFO[2019-12-13T15:19:04.302109943+09:00] loading plugin "io.containerd.gc.v1.scheduler"...  type=io.containerd.gc.v1
INFO[2019-12-13T15:19:04.302223446+09:00] loading plugin "io.containerd.service.v1.containers-service"...  type=io.containerd.service.v1
INFO[2019-12-13T15:19:04.302248170+09:00] loading plugin "io.containerd.service.v1.content-service"...  type=io.containerd.service.v1
INFO[2019-12-13T15:19:04.302372390+09:00] loading plugin "io.containerd.service.v1.diff-service"...  type=io.containerd.service.v1
INFO[2019-12-13T15:19:04.302397652+09:00] loading plugin "io.containerd.service.v1.images-service"...  type=io.containerd.service.v1
INFO[2019-12-13T15:19:04.302495566+09:00] loading plugin "io.containerd.service.v1.leases-service"...  type=io.containerd.service.v1
INFO[2019-12-13T15:19:04.302520976+09:00] loading plugin "io.containerd.service.v1.namespaces-service"...  type=io.containerd.service.v1
INFO[2019-12-13T15:19:04.302608714+09:00] loading plugin "io.containerd.service.v1.snapshots-service"...  type=io.containerd.service.v1
INFO[2019-12-13T15:19:04.302714671+09:00] loading plugin "io.containerd.runtime.v1.linux"...  type=io.containerd.runtime.v1
INFO[2019-12-13T15:19:04.302791529+09:00] loading plugin "io.containerd.runtime.v2.task"...  type=io.containerd.runtime.v2
INFO[2019-12-13T15:19:04.303051446+09:00] loading plugin "io.containerd.monitor.v1.cgroups"...  type=io.containerd.monitor.v1
INFO[2019-12-13T15:19:04.303622805+09:00] loading plugin "io.containerd.monitor.v1.cgroups-v2"...  type=io.containerd.monitor.v1
panic: descriptor Desc{fqName: "container_pids_current", help: "The current number of pids", constLabels: {}, variableLabels: [container_id namespace]} already exists with the same fully-qualified name and const label values

goroutine 1 [running]:
github.com/containerd/containerd/vendor/github.com/prometheus/client_golang/prometheus.(*Registry).MustRegister(0xc000085e80, 0xc00015b5b0, 0x1, 0x1)
        /home/suda/gopath/src/github.com/containerd/containerd/vendor/github.com/prometheus/client_golang/prometheus/registry.go:362 +0xaf
github.com/containerd/containerd/vendor/github.com/prometheus/client_golang/prometheus.MustRegister(...)
        /home/suda/gopath/src/github.com/containerd/containerd/vendor/github.com/prometheus/client_golang/prometheus/registry.go:154
github.com/containerd/containerd/vendor/github.com/docker/go-metrics.Register(...)
        /home/suda/gopath/src/github.com/containerd/containerd/vendor/github.com/docker/go-metrics/register.go:8
github.com/containerd/containerd/metrics/cgroups/v2.New(0xc00031ed80, 0x0, 0x0, 0x0, 0x0)
        /home/suda/gopath/src/github.com/containerd/containerd/metrics/cgroups/v2/cgroups.go:56 +0x361
github.com/containerd/containerd/plugin.(*Registration).Init(0xc0000879e0, 0xc00031ed80, 0x562c7f7716a0)
        /home/suda/gopath/src/github.com/containerd/containerd/plugin/plugin.go:110 +0x3a
github.com/containerd/containerd/services/server.New(0x562c7fa9cd60, 0xc000042098, 0xc000318f00, 0x1, 0x1, 0xc0003078c0)
        /home/suda/gopath/src/github.com/containerd/containerd/services/server/server.go:167 +0xcaa
github.com/containerd/containerd/cmd/containerd/command.App.func1(0xc0000d2580, 0x0, 0xc0002482e0)
        /home/suda/gopath/src/github.com/containerd/containerd/cmd/containerd/command/main.go:177 +0x7fa
github.com/containerd/containerd/vendor/github.com/urfave/cli.HandleAction(0x562c7f7e2180, 0x562c7fa3cf90, 0xc0000d2580, 0xc0000d2580, 0x0)
        /home/suda/gopath/src/github.com/containerd/containerd/vendor/github.com/urfave/cli/app.go:523 +0xc0
github.com/containerd/containerd/vendor/github.com/urfave/cli.(*App).Run(0xc000316c40, 0xc00003c1d0, 0x1, 0x1, 0x0, 0x0)
        /home/suda/gopath/src/github.com/containerd/containerd/vendor/github.com/urfave/cli/app.go:285 +0x5e1
main.main()
        github.com/containerd/containerd/cmd/containerd/main.go:33 +0x51
```

--- a/cmd/ctr/commands/tasks/metrics.go
+++ b/cmd/ctr/commands/tasks/metrics.go
@@ -152,6 +152,20 @@ func printCgroup2MetricsTable(w *tabwriter.Writer, data *v2.Metrics) {
 		fmt.Fprintf(w, "pids.current\t%v\t\n", data.Pids.Current)
 		fmt.Fprintf(w, "pids.limit\t%v\t\n", data.Pids.Limit)
 	}
+	if data.CPU != nil {
+		fmt.Fprintf(w, "cpu.usage_usec\t%v\t\n", data.CPU.UsageUsec)
+		fmt.Fprintf(w, "cpu.user_usec\t%v\t\n", data.CPU.UserUsec)
+		fmt.Fprintf(w, "cpu.sys_usec\t%v\t\n", data.CPU.SystemUsec)
+		fmt.Fprintf(w, "cpu.number_of_periods\t%v\t\n", data.CPU.NrPeriods)
+		fmt.Fprintf(w, "cpu.number_of_throttled\t%v\t\n", data.CPU.NrThrottled)
+		fmt.Fprintf(w, "cpu.throttled_usec\t%v\t\n", data.CPU.ThrottledUsec)
+	}
+	if data.Memory != nil {
+		fmt.Fprintf(w, "memory.usage\t%v\t\n", data.Memory.Usage)
+		fmt.Fprintf(w, "memory.usage_limit\t%v\t\n", data.Memory.UsageLimit)
+		fmt.Fprintf(w, "memory.swap\t%v\t\n", data.Memory.SwapUsage)
+		fmt.Fprintf(w, "memory.swap_limit\t%v\t\n", data.Memory.SwapLimit)
+	}
 }
 
 func printWindowsContainerStatistics(w *tabwriter.Writer, stats *wstats.WindowsContainerStatistics) {

--- a/cmd/ctr/commands/tasks/metrics.go
+++ b/cmd/ctr/commands/tasks/metrics.go
@@ -163,7 +163,7 @@ func printCgroup2MetricsTable(w *tabwriter.Writer, data *v2.Metrics) {
 	if data.Memory != nil {
 		fmt.Fprintf(w, "memory.usage\t%v\t\n", data.Memory.Usage)
 		fmt.Fprintf(w, "memory.usage_limit\t%v\t\n", data.Memory.UsageLimit)
-		fmt.Fprintf(w, "memory.swap\t%v\t\n", data.Memory.SwapUsage)
+		fmt.Fprintf(w, "memory.swap_usage\t%v\t\n", data.Memory.SwapUsage)
 		fmt.Fprintf(w, "memory.swap_limit\t%v\t\n", data.Memory.SwapLimit)
 	}
 }

--- a/cmd/ctr/commands/tasks/metrics.go
+++ b/cmd/ctr/commands/tasks/metrics.go
@@ -155,9 +155,9 @@ func printCgroup2MetricsTable(w *tabwriter.Writer, data *v2.Metrics) {
 	if data.CPU != nil {
 		fmt.Fprintf(w, "cpu.usage_usec\t%v\t\n", data.CPU.UsageUsec)
 		fmt.Fprintf(w, "cpu.user_usec\t%v\t\n", data.CPU.UserUsec)
-		fmt.Fprintf(w, "cpu.sys_usec\t%v\t\n", data.CPU.SystemUsec)
-		fmt.Fprintf(w, "cpu.number_of_periods\t%v\t\n", data.CPU.NrPeriods)
-		fmt.Fprintf(w, "cpu.number_of_throttled\t%v\t\n", data.CPU.NrThrottled)
+		fmt.Fprintf(w, "cpu.system_usec\t%v\t\n", data.CPU.SystemUsec)
+		fmt.Fprintf(w, "cpu.nr_periods\t%v\t\n", data.CPU.NrPeriods)
+		fmt.Fprintf(w, "cpu.nr_throttled\t%v\t\n", data.CPU.NrThrottled)
 		fmt.Fprintf(w, "cpu.throttled_usec\t%v\t\n", data.CPU.ThrottledUsec)
 	}
 	if data.Memory != nil {

--- a/metrics/cgroups/v2/cpu.go
+++ b/metrics/cgroups/v2/cpu.go
@@ -1,0 +1,124 @@
+// +build linux
+
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package v2
+
+import (
+	v2 "github.com/containerd/containerd/metrics/types/v2"
+	metrics "github.com/docker/go-metrics"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+var cpuMetrics = []*metric{
+	{
+		name: "cpu",
+		help: "Current cpu usage_usec (cgroup v2)",
+		unit: metrics.Unit("usage_usec"),
+		vt:   prometheus.GaugeValue,
+		getValues: func(stats *v2.Metrics) []value {
+			if stats.CPU == nil {
+				return nil
+			}
+			return []value{
+				{
+					v: float64(stats.CPU.UsageUsec),
+				},
+			}
+		},
+	},
+	{
+		name: "cpu",
+		help: "Current cpu user_usec (cgroup v2)",
+		unit: metrics.Unit("user_usec"),
+		vt:   prometheus.GaugeValue,
+		getValues: func(stats *v2.Metrics) []value {
+			if stats.CPU == nil {
+				return nil
+			}
+			return []value{
+				{
+					v: float64(stats.CPU.UserUsec),
+				},
+			}
+		},
+	},
+	{
+		name: "cpu",
+		help: "Current cpu system_usec (cgroup v2)",
+		unit: metrics.Unit("system_usec"),
+		vt:   prometheus.GaugeValue,
+		getValues: func(stats *v2.Metrics) []value {
+			if stats.CPU == nil {
+				return nil
+			}
+			return []value{
+				{
+					v: float64(stats.CPU.SystemUsec),
+				},
+			}
+		},
+	},
+	{
+		name: "cpu",
+		help: "Current cpu nr_periods (only if controller is enabled)",
+		unit: metrics.Unit("nr_periods"),
+		vt:   prometheus.GaugeValue,
+		getValues: func(stats *v2.Metrics) []value {
+			if stats.CPU == nil {
+				return nil
+			}
+			return []value{
+				{
+					v: float64(stats.CPU.NrPeriods),
+				},
+			}
+		},
+	},
+	{
+		name: "cpu",
+		help: "Current cpu nr_throttled (only if controller is enabled)",
+		unit: metrics.Unit("nr_throttled"),
+		vt:   prometheus.GaugeValue,
+		getValues: func(stats *v2.Metrics) []value {
+			if stats.CPU == nil {
+				return nil
+			}
+			return []value{
+				{
+					v: float64(stats.CPU.NrThrottled),
+				},
+			}
+		},
+	},
+	{
+		name: "cpu",
+		help: "Current cpu throttled_usec (only if controller is enabled)",
+		unit: metrics.Unit("throttled_usec"),
+		vt:   prometheus.GaugeValue,
+		getValues: func(stats *v2.Metrics) []value {
+			if stats.CPU == nil {
+				return nil
+			}
+			return []value{
+				{
+					v: float64(stats.CPU.ThrottledUsec),
+				},
+			}
+		},
+	},
+}

--- a/metrics/cgroups/v2/cpu.go
+++ b/metrics/cgroups/v2/cpu.go
@@ -28,7 +28,7 @@ var cpuMetrics = []*metric{
 	{
 		name: "cpu_usage",
 		help: "Total cpu usage (cgroup v2)",
-		unit: metrics.Nanoseconds,
+		unit: metrics.Unit("microseconds"),
 		vt:   prometheus.GaugeValue,
 		getValues: func(stats *v2.Metrics) []value {
 			if stats.CPU == nil {
@@ -44,7 +44,7 @@ var cpuMetrics = []*metric{
 	{
 		name: "cpu_user_usage",
 		help: "Current cpu usage in user space (cgroup v2)",
-		unit: metrics.Nanoseconds,
+		unit: metrics.Unit("microseconds"),
 		vt:   prometheus.GaugeValue,
 		getValues: func(stats *v2.Metrics) []value {
 			if stats.CPU == nil {
@@ -60,7 +60,7 @@ var cpuMetrics = []*metric{
 	{
 		name: "cpu_kernel_usage",
 		help: "Current cpu usage in kernel space (cgroup v2)",
-		unit: metrics.Nanoseconds,
+		unit: metrics.Unit("microseconds"),
 		vt:   prometheus.GaugeValue,
 		getValues: func(stats *v2.Metrics) []value {
 			if stats.CPU == nil {
@@ -108,7 +108,7 @@ var cpuMetrics = []*metric{
 	{
 		name: "cpu_throttled_usec",
 		help: "Total time duration for which tasks have been throttled. (only if controller is enabled)",
-		unit: metrics.Nanoseconds,
+		unit: metrics.Unit("microseconds"),
 		vt:   prometheus.GaugeValue,
 		getValues: func(stats *v2.Metrics) []value {
 			if stats.CPU == nil {

--- a/metrics/cgroups/v2/cpu.go
+++ b/metrics/cgroups/v2/cpu.go
@@ -26,9 +26,9 @@ import (
 
 var cpuMetrics = []*metric{
 	{
-		name: "cpu",
-		help: "Current cpu usage_usec (cgroup v2)",
-		unit: metrics.Unit("usage_usec"),
+		name: "cpu_usage",
+		help: "Total cpu usage (cgroup v2)",
+		unit: metrics.Nanoseconds,
 		vt:   prometheus.GaugeValue,
 		getValues: func(stats *v2.Metrics) []value {
 			if stats.CPU == nil {
@@ -42,9 +42,9 @@ var cpuMetrics = []*metric{
 		},
 	},
 	{
-		name: "cpu",
-		help: "Current cpu user_usec (cgroup v2)",
-		unit: metrics.Unit("user_usec"),
+		name: "cpu_user_usage",
+		help: "Current cpu usage in user space (cgroup v2)",
+		unit: metrics.Nanoseconds,
 		vt:   prometheus.GaugeValue,
 		getValues: func(stats *v2.Metrics) []value {
 			if stats.CPU == nil {
@@ -58,9 +58,9 @@ var cpuMetrics = []*metric{
 		},
 	},
 	{
-		name: "cpu",
-		help: "Current cpu system_usec (cgroup v2)",
-		unit: metrics.Unit("system_usec"),
+		name: "cpu_kernel_usage",
+		help: "Current cpu usage in kernel space (cgroup v2)",
+		unit: metrics.Nanoseconds,
 		vt:   prometheus.GaugeValue,
 		getValues: func(stats *v2.Metrics) []value {
 			if stats.CPU == nil {
@@ -74,9 +74,9 @@ var cpuMetrics = []*metric{
 		},
 	},
 	{
-		name: "cpu",
-		help: "Current cpu nr_periods (only if controller is enabled)",
-		unit: metrics.Unit("nr_periods"),
+		name: "cpu_nr_periods",
+		help: "Current cpu number of periods (only if controller is enabled)",
+		unit: metrics.Total,
 		vt:   prometheus.GaugeValue,
 		getValues: func(stats *v2.Metrics) []value {
 			if stats.CPU == nil {
@@ -90,9 +90,9 @@ var cpuMetrics = []*metric{
 		},
 	},
 	{
-		name: "cpu",
-		help: "Current cpu nr_throttled (only if controller is enabled)",
-		unit: metrics.Unit("nr_throttled"),
+		name: "cpu_nr_throttled",
+		help: "Total number of times tasks have been throttled (only if controller is enabled)",
+		unit: metrics.Total,
 		vt:   prometheus.GaugeValue,
 		getValues: func(stats *v2.Metrics) []value {
 			if stats.CPU == nil {
@@ -106,9 +106,9 @@ var cpuMetrics = []*metric{
 		},
 	},
 	{
-		name: "cpu",
-		help: "Current cpu throttled_usec (only if controller is enabled)",
-		unit: metrics.Unit("throttled_usec"),
+		name: "cpu_throttled_usec",
+		help: "Total time duration for which tasks have been throttled. (only if controller is enabled)",
+		unit: metrics.Nanoseconds,
 		vt:   prometheus.GaugeValue,
 		getValues: func(stats *v2.Metrics) []value {
 			if stats.CPU == nil {

--- a/metrics/cgroups/v2/cpu.go
+++ b/metrics/cgroups/v2/cpu.go
@@ -26,7 +26,7 @@ import (
 
 var cpuMetrics = []*metric{
 	{
-		name: "cpu_usage",
+		name: "cpu_usage_usec",
 		help: "Total cpu usage (cgroup v2)",
 		unit: metrics.Unit("microseconds"),
 		vt:   prometheus.GaugeValue,
@@ -42,7 +42,7 @@ var cpuMetrics = []*metric{
 		},
 	},
 	{
-		name: "cpu_user_usage",
+		name: "cpu_user_usec",
 		help: "Current cpu usage in user space (cgroup v2)",
 		unit: metrics.Unit("microseconds"),
 		vt:   prometheus.GaugeValue,
@@ -58,7 +58,7 @@ var cpuMetrics = []*metric{
 		},
 	},
 	{
-		name: "cpu_system_usage",
+		name: "cpu_system_usec",
 		help: "Current cpu usage in kernel space (cgroup v2)",
 		unit: metrics.Unit("microseconds"),
 		vt:   prometheus.GaugeValue,

--- a/metrics/cgroups/v2/cpu.go
+++ b/metrics/cgroups/v2/cpu.go
@@ -58,7 +58,7 @@ var cpuMetrics = []*metric{
 		},
 	},
 	{
-		name: "cpu_kernel_usage",
+		name: "cpu_system_usage",
 		help: "Current cpu usage in kernel space (cgroup v2)",
 		unit: metrics.Unit("microseconds"),
 		vt:   prometheus.GaugeValue,

--- a/metrics/cgroups/v2/memory.go
+++ b/metrics/cgroups/v2/memory.go
@@ -91,8 +91,8 @@ var memoryMetrics = []*metric{
 	},
 
 	{
-		name: "memory_mapped_file",
-		help: "The mapped_file amount used",
+		name: "memory_file_mapped",
+		help: "The file_mapped amount",
 		unit: metrics.Bytes,
 		vt:   prometheus.GaugeValue,
 		getValues: func(stats *v2.Metrics) []value {
@@ -107,8 +107,8 @@ var memoryMetrics = []*metric{
 		},
 	},
 	{
-		name: "memory_dirty_file",
-		help: "The dirty file amount",
+		name: "memory_file_dirty",
+		help: "The file_dirty amount",
 		unit: metrics.Bytes,
 		vt:   prometheus.GaugeValue,
 		getValues: func(stats *v2.Metrics) []value {

--- a/metrics/cgroups/v2/memory.go
+++ b/metrics/cgroups/v2/memory.go
@@ -1,0 +1,92 @@
+// +build linux
+
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package v2
+
+import (
+	v2 "github.com/containerd/containerd/metrics/types/v2"
+	metrics "github.com/docker/go-metrics"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+var memoryMetrics = []*metric{
+	{
+		name: "memory",
+		help: "Current memory usage (cgroup v2)",
+		unit: metrics.Unit("usage"),
+		vt:   prometheus.GaugeValue,
+		getValues: func(stats *v2.Metrics) []value {
+			if stats.Memory == nil {
+				return nil
+			}
+			return []value{
+				{
+					v: float64(stats.Memory.Usage),
+				},
+			}
+		},
+	},
+	{
+		name: "memory",
+		help: "Current memory usage limit (cgroup v2)",
+		unit: metrics.Unit("limit"),
+		vt:   prometheus.GaugeValue,
+		getValues: func(stats *v2.Metrics) []value {
+			if stats.Memory == nil {
+				return nil
+			}
+			return []value{
+				{
+					v: float64(stats.Memory.UsageLimit),
+				},
+			}
+		},
+	},
+	{
+		name: "memory",
+		help: "Current swap usage (cgroup v2)",
+		unit: metrics.Unit("swap"),
+		vt:   prometheus.GaugeValue,
+		getValues: func(stats *v2.Metrics) []value {
+			if stats.Memory == nil {
+				return nil
+			}
+			return []value{
+				{
+					v: float64(stats.Memory.SwapUsage),
+				},
+			}
+		},
+	},
+	{
+		name: "memory",
+		help: "Current swap usage limit (cgroup v2)",
+		unit: metrics.Unit("swap_limit"),
+		vt:   prometheus.GaugeValue,
+		getValues: func(stats *v2.Metrics) []value {
+			if stats.Memory == nil {
+				return nil
+			}
+			return []value{
+				{
+					v: float64(stats.Memory.SwapLimit),
+				},
+			}
+		},
+	},
+}

--- a/metrics/cgroups/v2/memory.go
+++ b/metrics/cgroups/v2/memory.go
@@ -74,7 +74,7 @@ var memoryMetrics = []*metric{
 		},
 	},
 	{
-		name: "memory_swap_Limit",
+		name: "memory_swap_limit",
 		help: "Current swap usage limit (cgroup v2)",
 		unit: metrics.Bytes,
 		vt:   prometheus.GaugeValue,

--- a/metrics/cgroups/v2/memory.go
+++ b/metrics/cgroups/v2/memory.go
@@ -26,9 +26,9 @@ import (
 
 var memoryMetrics = []*metric{
 	{
-		name: "memory",
+		name: "memory_usage",
 		help: "Current memory usage (cgroup v2)",
-		unit: metrics.Unit("usage"),
+		unit: metrics.Bytes,
 		vt:   prometheus.GaugeValue,
 		getValues: func(stats *v2.Metrics) []value {
 			if stats.Memory == nil {
@@ -42,9 +42,9 @@ var memoryMetrics = []*metric{
 		},
 	},
 	{
-		name: "memory",
+		name: "memory_limit",
 		help: "Current memory usage limit (cgroup v2)",
-		unit: metrics.Unit("limit"),
+		unit: metrics.Bytes,
 		vt:   prometheus.GaugeValue,
 		getValues: func(stats *v2.Metrics) []value {
 			if stats.Memory == nil {
@@ -58,9 +58,9 @@ var memoryMetrics = []*metric{
 		},
 	},
 	{
-		name: "memory",
+		name: "memory_swap_usage",
 		help: "Current swap usage (cgroup v2)",
-		unit: metrics.Unit("swap"),
+		unit: metrics.Bytes,
 		vt:   prometheus.GaugeValue,
 		getValues: func(stats *v2.Metrics) []value {
 			if stats.Memory == nil {
@@ -74,9 +74,9 @@ var memoryMetrics = []*metric{
 		},
 	},
 	{
-		name: "memory",
+		name: "memory_swap_Limit",
 		help: "Current swap usage limit (cgroup v2)",
-		unit: metrics.Unit("swap_limit"),
+		unit: metrics.Bytes,
 		vt:   prometheus.GaugeValue,
 		getValues: func(stats *v2.Metrics) []value {
 			if stats.Memory == nil {
@@ -85,6 +85,503 @@ var memoryMetrics = []*metric{
 			return []value{
 				{
 					v: float64(stats.Memory.SwapLimit),
+				},
+			}
+		},
+	},
+
+	{
+		name: "memory_mapped_file",
+		help: "The mapped_file amount used",
+		unit: metrics.Bytes,
+		vt:   prometheus.GaugeValue,
+		getValues: func(stats *v2.Metrics) []value {
+			if stats.Memory == nil {
+				return nil
+			}
+			return []value{
+				{
+					v: float64(stats.Memory.FileMapped),
+				},
+			}
+		},
+	},
+	{
+		name: "memory_dirty_file",
+		help: "The dirty file amount",
+		unit: metrics.Bytes,
+		vt:   prometheus.GaugeValue,
+		getValues: func(stats *v2.Metrics) []value {
+			if stats.Memory == nil {
+				return nil
+			}
+			return []value{
+				{
+					v: float64(stats.Memory.FileDirty),
+				},
+			}
+		},
+	},
+	{
+		name: "memory_file_writeback",
+		help: "The file_writeback amount",
+		unit: metrics.Bytes,
+		vt:   prometheus.GaugeValue,
+		getValues: func(stats *v2.Metrics) []value {
+			if stats.Memory == nil {
+				return nil
+			}
+			return []value{
+				{
+					v: float64(stats.Memory.FileWriteback),
+				},
+			}
+		},
+	},
+	{
+		name: "memory_pgactive",
+		help: "The pgactive amount",
+		unit: metrics.Bytes,
+		vt:   prometheus.GaugeValue,
+		getValues: func(stats *v2.Metrics) []value {
+			if stats.Memory == nil {
+				return nil
+			}
+			return []value{
+				{
+					v: float64(stats.Memory.Pgactivate),
+				},
+			}
+		},
+	},
+	{
+		name: "memory_pgdeactivate",
+		help: "The pgdeactivate amount",
+		unit: metrics.Bytes,
+		vt:   prometheus.GaugeValue,
+		getValues: func(stats *v2.Metrics) []value {
+			if stats.Memory == nil {
+				return nil
+			}
+			return []value{
+				{
+					v: float64(stats.Memory.Pgdeactivate),
+				},
+			}
+		},
+	},
+	{
+		name: "memory_pgfault",
+		help: "The pgfault amount",
+		unit: metrics.Bytes,
+		vt:   prometheus.GaugeValue,
+		getValues: func(stats *v2.Metrics) []value {
+			if stats.Memory == nil {
+				return nil
+			}
+			return []value{
+				{
+					v: float64(stats.Memory.Pgfault),
+				},
+			}
+		},
+	},
+	{
+		name: "memory_pgmajfault",
+		help: "The pgmajfault amount",
+		unit: metrics.Bytes,
+		vt:   prometheus.GaugeValue,
+		getValues: func(stats *v2.Metrics) []value {
+			if stats.Memory == nil {
+				return nil
+			}
+			return []value{
+				{
+					v: float64(stats.Memory.Pgmajfault),
+				},
+			}
+		},
+	},
+	{
+		name: "memory_pglazyfree",
+		help: "The pglazyfree amount",
+		unit: metrics.Bytes,
+		vt:   prometheus.GaugeValue,
+		getValues: func(stats *v2.Metrics) []value {
+			if stats.Memory == nil {
+				return nil
+			}
+			return []value{
+				{
+					v: float64(stats.Memory.Pglazyfree),
+				},
+			}
+		},
+	},
+	{
+		name: "memory_pgrefill",
+		help: "The pgrefill amount",
+		unit: metrics.Bytes,
+		vt:   prometheus.GaugeValue,
+		getValues: func(stats *v2.Metrics) []value {
+			if stats.Memory == nil {
+				return nil
+			}
+			return []value{
+				{
+					v: float64(stats.Memory.Pgrefill),
+				},
+			}
+		},
+	},
+	{
+		name: "memory_pglazyfreed",
+		help: "The pglazyfreed amount",
+		unit: metrics.Bytes,
+		vt:   prometheus.GaugeValue,
+		getValues: func(stats *v2.Metrics) []value {
+			if stats.Memory == nil {
+				return nil
+			}
+			return []value{
+				{
+					v: float64(stats.Memory.Pglazyfreed),
+				},
+			}
+		},
+	},
+	{
+		name: "memory_pgscan",
+		help: "The pgscan amount",
+		unit: metrics.Bytes,
+		vt:   prometheus.GaugeValue,
+		getValues: func(stats *v2.Metrics) []value {
+			if stats.Memory == nil {
+				return nil
+			}
+			return []value{
+				{
+					v: float64(stats.Memory.Pgscan),
+				},
+			}
+		},
+	},
+	{
+		name: "memory_pgsteal",
+		help: "The pgsteal amount",
+		unit: metrics.Bytes,
+		vt:   prometheus.GaugeValue,
+		getValues: func(stats *v2.Metrics) []value {
+			if stats.Memory == nil {
+				return nil
+			}
+			return []value{
+				{
+					v: float64(stats.Memory.Pgsteal),
+				},
+			}
+		},
+	},
+	{
+		name: "memory_inactive_anon",
+		help: "The inactive_anon amount",
+		unit: metrics.Bytes,
+		vt:   prometheus.GaugeValue,
+		getValues: func(stats *v2.Metrics) []value {
+			if stats.Memory == nil {
+				return nil
+			}
+			return []value{
+				{
+					v: float64(stats.Memory.InactiveAnon),
+				},
+			}
+		},
+	},
+	{
+		name: "memory_active_anon",
+		help: "The active_anon amount",
+		unit: metrics.Bytes,
+		vt:   prometheus.GaugeValue,
+		getValues: func(stats *v2.Metrics) []value {
+			if stats.Memory == nil {
+				return nil
+			}
+			return []value{
+				{
+					v: float64(stats.Memory.ActiveAnon),
+				},
+			}
+		},
+	},
+	{
+		name: "memory_inactive_file",
+		help: "The inactive_file amount",
+		unit: metrics.Bytes,
+		vt:   prometheus.GaugeValue,
+		getValues: func(stats *v2.Metrics) []value {
+			if stats.Memory == nil {
+				return nil
+			}
+			return []value{
+				{
+					v: float64(stats.Memory.InactiveFile),
+				},
+			}
+		},
+	},
+	{
+		name: "memory_active_file",
+		help: "The active_file amount",
+		unit: metrics.Bytes,
+		vt:   prometheus.GaugeValue,
+		getValues: func(stats *v2.Metrics) []value {
+			if stats.Memory == nil {
+				return nil
+			}
+			return []value{
+				{
+					v: float64(stats.Memory.ActiveFile),
+				},
+			}
+		},
+	},
+	{
+		name: "memory_unevictable",
+		help: "The unevictable amount",
+		unit: metrics.Bytes,
+		vt:   prometheus.GaugeValue,
+		getValues: func(stats *v2.Metrics) []value {
+			if stats.Memory == nil {
+				return nil
+			}
+			return []value{
+				{
+					v: float64(stats.Memory.Unevictable),
+				},
+			}
+		},
+	},
+	{
+		name: "memory_anon",
+		help: "The anon amount",
+		unit: metrics.Bytes,
+		vt:   prometheus.GaugeValue,
+		getValues: func(stats *v2.Metrics) []value {
+			if stats.Memory == nil {
+				return nil
+			}
+			return []value{
+				{
+					v: float64(stats.Memory.Anon),
+				},
+			}
+		},
+	},
+	{
+		name: "memory_file",
+		help: "The file amount",
+		unit: metrics.Bytes,
+		vt:   prometheus.GaugeValue,
+		getValues: func(stats *v2.Metrics) []value {
+			if stats.Memory == nil {
+				return nil
+			}
+			return []value{
+				{
+					v: float64(stats.Memory.File),
+				},
+			}
+		},
+	},
+	{
+		name: "memory_kernel_stack",
+		help: "The kernel_stack amount",
+		unit: metrics.Bytes,
+		vt:   prometheus.GaugeValue,
+		getValues: func(stats *v2.Metrics) []value {
+			if stats.Memory == nil {
+				return nil
+			}
+			return []value{
+				{
+					v: float64(stats.Memory.KernelStack),
+				},
+			}
+		},
+	},
+	{
+		name: "memory_slab",
+		help: "The slab amount",
+		unit: metrics.Bytes,
+		vt:   prometheus.GaugeValue,
+		getValues: func(stats *v2.Metrics) []value {
+			if stats.Memory == nil {
+				return nil
+			}
+			return []value{
+				{
+					v: float64(stats.Memory.Slab),
+				},
+			}
+		},
+	},
+	{
+		name: "memory_sock",
+		help: "The sock amount",
+		unit: metrics.Bytes,
+		vt:   prometheus.GaugeValue,
+		getValues: func(stats *v2.Metrics) []value {
+			if stats.Memory == nil {
+				return nil
+			}
+			return []value{
+				{
+					v: float64(stats.Memory.Sock),
+				},
+			}
+		},
+	},
+	{
+		name: "memory_shmem",
+		help: "The shmem amount",
+		unit: metrics.Bytes,
+		vt:   prometheus.GaugeValue,
+		getValues: func(stats *v2.Metrics) []value {
+			if stats.Memory == nil {
+				return nil
+			}
+			return []value{
+				{
+					v: float64(stats.Memory.Shmem),
+				},
+			}
+		},
+	},
+	{
+		name: "memory_anon_thp",
+		help: "The anon_thp amount",
+		unit: metrics.Bytes,
+		vt:   prometheus.GaugeValue,
+		getValues: func(stats *v2.Metrics) []value {
+			if stats.Memory == nil {
+				return nil
+			}
+			return []value{
+				{
+					v: float64(stats.Memory.AnonThp),
+				},
+			}
+		},
+	},
+	{
+		name: "memory_slab_reclaimable",
+		help: "The slab_reclaimable amount",
+		unit: metrics.Bytes,
+		vt:   prometheus.GaugeValue,
+		getValues: func(stats *v2.Metrics) []value {
+			if stats.Memory == nil {
+				return nil
+			}
+			return []value{
+				{
+					v: float64(stats.Memory.SlabReclaimable),
+				},
+			}
+		},
+	},
+	{
+		name: "memory_slab_unreclaimable",
+		help: "The slab_unreclaimable amount",
+		unit: metrics.Bytes,
+		vt:   prometheus.GaugeValue,
+		getValues: func(stats *v2.Metrics) []value {
+			if stats.Memory == nil {
+				return nil
+			}
+			return []value{
+				{
+					v: float64(stats.Memory.SlabUnreclaimable),
+				},
+			}
+		},
+	},
+	{
+		name: "memory_workingset_refault",
+		help: "The workingset_refault amount",
+		unit: metrics.Bytes,
+		vt:   prometheus.GaugeValue,
+		getValues: func(stats *v2.Metrics) []value {
+			if stats.Memory == nil {
+				return nil
+			}
+			return []value{
+				{
+					v: float64(stats.Memory.WorkingsetRefault),
+				},
+			}
+		},
+	},
+	{
+		name: "memory_workingset_activate",
+		help: "The workingset_activate amount",
+		unit: metrics.Bytes,
+		vt:   prometheus.GaugeValue,
+		getValues: func(stats *v2.Metrics) []value {
+			if stats.Memory == nil {
+				return nil
+			}
+			return []value{
+				{
+					v: float64(stats.Memory.WorkingsetActivate),
+				},
+			}
+		},
+	},
+	{
+		name: "memory_workingset_nodereclaim",
+		help: "The workingset_nodereclaim amount",
+		unit: metrics.Bytes,
+		vt:   prometheus.GaugeValue,
+		getValues: func(stats *v2.Metrics) []value {
+			if stats.Memory == nil {
+				return nil
+			}
+			return []value{
+				{
+					v: float64(stats.Memory.WorkingsetNodereclaim),
+				},
+			}
+		},
+	},
+	{
+		name: "memory_thp_fault_alloc",
+		help: "The thp_fault_alloc amount",
+		unit: metrics.Bytes,
+		vt:   prometheus.GaugeValue,
+		getValues: func(stats *v2.Metrics) []value {
+			if stats.Memory == nil {
+				return nil
+			}
+			return []value{
+				{
+					v: float64(stats.Memory.ThpFaultAlloc),
+				},
+			}
+		},
+	},
+	{
+		name: "memory_thp_collapse_alloc",
+		help: "The thp_collapse_alloc amount",
+		unit: metrics.Bytes,
+		vt:   prometheus.GaugeValue,
+		getValues: func(stats *v2.Metrics) []value {
+			if stats.Memory == nil {
+				return nil
+			}
+			return []value{
+				{
+					v: float64(stats.Memory.ThpCollapseAlloc),
 				},
 			}
 		},

--- a/metrics/cgroups/v2/memory.go
+++ b/metrics/cgroups/v2/memory.go
@@ -139,8 +139,8 @@ var memoryMetrics = []*metric{
 		},
 	},
 	{
-		name: "memory_pgactive",
-		help: "The pgactive amount",
+		name: "memory_pgactivate",
+		help: "The pgactivate amount",
 		unit: metrics.Bytes,
 		vt:   prometheus.GaugeValue,
 		getValues: func(stats *v2.Metrics) []value {

--- a/metrics/cgroups/v2/memory.go
+++ b/metrics/cgroups/v2/memory.go
@@ -42,7 +42,7 @@ var memoryMetrics = []*metric{
 		},
 	},
 	{
-		name: "memory_limit",
+		name: "memory_usage_limit",
 		help: "Current memory usage limit (cgroup v2)",
 		unit: metrics.Bytes,
 		vt:   prometheus.GaugeValue,

--- a/metrics/cgroups/v2/metrics.go
+++ b/metrics/cgroups/v2/metrics.go
@@ -43,6 +43,8 @@ func newCollector(ns *metrics.Namespace) *collector {
 		tasks: make(map[string]runtime.Task),
 	}
 	c.metrics = append(c.metrics, pidMetrics...)
+	c.metrics = append(c.metrics, cpuMetrics...)
+	c.metrics = append(c.metrics, memoryMetrics...)
 	c.storedMetrics = make(chan prometheus.Metric, 100*len(c.metrics))
 	ns.Add(c)
 	return c

